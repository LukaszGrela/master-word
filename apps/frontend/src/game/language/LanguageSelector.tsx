import { FC, useCallback } from 'react';
import { IProps } from './types';
import { classNames } from '@repo/utils';
import { TSupportedLanguages } from '../../api';
import { useLanguage } from '../../i18n';

import './LanguageSelector.scss';

const LanguageSelector: FC<IProps> = ({
  language,
  onClick,
  className,
  translationOverride,
}): JSX.Element => {
  const { getUIText: t } = useLanguage();

  const handleLanguageChange = useCallback(
    (selected: TSupportedLanguages): void => {
      if (language !== selected) {
        onClick(selected);
      }
    },
    [language, onClick],
  );
  return (
    <div className={classNames('language-selector', 'translation', className)}>
      <span className="hidden sr">
        {translationOverride?.screenReaderInfo ?? t('translation-info-sr')}
      </span>
      <button
        title={
          translationOverride?.buttonTitles?.pl ??
          t('translation-button-polish')
        }
        className={classNames(
          'translation-btn',
          language === 'pl' && 'selected',
        )}
        onClick={() => {
          handleLanguageChange('pl');
        }}
      >
        {translationOverride?.buttonLabels?.pl && (
          <span className="label">{translationOverride?.buttonLabels?.pl}</span>
        )}
        <span className="flag">ðŸ‡µðŸ‡±</span>
      </button>
      <button
        title={
          translationOverride?.buttonTitles?.en ??
          t('translation-button-english')
        }
        className={classNames(
          'translation-btn',
          language === 'en' && 'selected',
        )}
        onClick={() => {
          handleLanguageChange('en');
        }}
      >
        {translationOverride?.buttonLabels?.en && (
          <span className="label">{translationOverride?.buttonLabels?.en}</span>
        )}
        <span className="flag">ðŸ‡ºðŸ‡¸</span>
      </button>
    </div>
  );
};

export default LanguageSelector;
