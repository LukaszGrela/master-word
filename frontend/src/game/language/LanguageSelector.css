.language-selector.translation {
  position: absolute;
  top: 0.25rem;
  right: 0.25rem;
  display: flex;
  flex-direction: row;
}

.language-selector.translation button {
  padding: 0.25rem;
}

.language-selector.translation button.selected {
  background-color: hsl(25 40% 90% / 1);
  border-color: hsl(27, 27%, 50%);
}

.language-selector.translation button:hover {
  border-color: hsl(27, 27%, 50%);
}
