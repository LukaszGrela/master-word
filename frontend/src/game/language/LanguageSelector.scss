@import '../../styles/variables';

.language-selector.translation {
  position: absolute;
  top: 0.25rem;
  right: 0.75rem;
  display: flex;
  flex-direction: row;
  gap: 0.25rem;

  & button {
    padding: 0.25rem;
    background-color: $language-btn-background;
    border-color: $language-btn-border;

    @media (prefers-color-scheme: dark) {
      background-color: $language-btn-background-dark;
    }
    &:hover {
      border-color: $language-btn-border-hover;
    }
    &.selected {
      background-color: $language-btn-selected-background;
      border-color: $language-btn-selected-border;
      @media (prefers-color-scheme: dark) {
        background-color: $language-btn-selected-background-dark;
      }
    }
  }
}
