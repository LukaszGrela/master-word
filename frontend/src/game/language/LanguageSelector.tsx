import { FC, useCallback } from 'react';
import { IProps } from './types';
import t from '../../i18n';
import { classNames } from '../../utils/classNames';
import { TSupportedLanguages } from '../../api';
import './LanguageSelector.css';

const LanguageSelector: FC<IProps> = ({
  language,
  onClick,
  className,
  translationOverride,
}): JSX.Element => {
  const handleLanguageChange = useCallback(
    (selected: TSupportedLanguages): void => {
      if (language !== selected) {
        onClick(selected);
      }
    },
    [language, onClick]
  );
  return (
    <div className={classNames('language-selector', 'translation', className)}>
      <span className='hidden sr'>
        {translationOverride?.screenReaderInfo || t('translation-info-sr')}
      </span>
      <button
        title={
          translationOverride?.buttonTitles?.pl ||
          t('translation-button-polish')
        }
        className={classNames(
          'translation-btn',
          language === 'pl' && 'selected'
        )}
        onClick={() => {
          handleLanguageChange('pl');
        }}
      >
        ðŸ‡µðŸ‡±
      </button>
      <button
        title={
          translationOverride?.buttonTitles?.en ||
          t('translation-button-english')
        }
        className={classNames(
          'translation-btn',
          language === 'en' && 'selected'
        )}
        onClick={() => {
          handleLanguageChange('en');
        }}
      >
        ðŸ‡ºðŸ‡¸
      </button>
    </div>
  );
};

export default LanguageSelector;
