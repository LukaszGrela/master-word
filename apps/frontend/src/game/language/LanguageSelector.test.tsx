import { describe, expect, it, vi } from 'vitest';
import { fireEvent } from '@testing-library/react';
import {
  contextRenderer,
  defaultLanguageContext,
} from '../../i18n/LanguageContext/__tests__/contextRenderer';

import LanguageSelector from './LanguageSelector';

describe('game', () => {
  describe('language', () => {
    describe('LanguageSelector', () => {
      it('Renders with required props', () => {
        const { container } = contextRenderer(
          <LanguageSelector language="pl" onClick={vi.fn()} list={[]} />,
        );

        expect(container).toMatchSnapshot();
      });
      it('Renders with context and required props', () => {
        const { container } = contextRenderer(
          <LanguageSelector language="pl" onClick={vi.fn()} list={[]} />,
          {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: vi.fn().mockImplementation(() => {
                return 'screen reader info';
              }),
            },
            renderOptions: {},
          },
        );

        expect(container).toMatchSnapshot();
      });
      it('Renders with context and all props', () => {
        const { container } = contextRenderer(
          <LanguageSelector
            language="pl"
            onClick={vi.fn()}
            list={[]}
            className="test"
            translationOverride={{
              screenReaderInfo: 'screenReaderInfo',
            }}
          />,
          {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: vi.fn().mockImplementation(() => {
                return 'screen reader info';
              }),
            },
            renderOptions: {},
          },
        );

        expect(container).toMatchSnapshot();
      });
      it('renders options', () => {
        const { container } = contextRenderer(
          <LanguageSelector
            language="pl"
            onClick={vi.fn()}
            list={[
              {
                title: 'Option 1 Button Title',
                value: 'option-1',
                icon: <>ICO</>,
                label: 'Option 1',
              },

              {
                title: 'Option 2 Button Title',
                value: 'option-2',
                icon: <>ICO</>,
                label: 'Option 2',
              },
            ]}
          />,
          {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: vi.fn().mockImplementation(() => {
                return 'screen reader info';
              }),
            },
            renderOptions: {},
          },
        );

        expect(container).toMatchSnapshot();
      });
      it('Calls onClick on rendered option', () => {
        const clickHandler = vi.fn();
        const list = [
          {
            title: 'Option 1 Button Title',
            value: 'pl',
            icon: <>ICO</>,
            label: 'Option 1',
          },

          {
            title: 'Option 2 Button Title',
            value: 'en',
            icon: <>ICO</>,
            label: 'Option 2',
          },
        ];
        const { getByTitle } = contextRenderer(
          <LanguageSelector language="pl" onClick={clickHandler} list={list} />,
          {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: vi.fn().mockImplementation(() => {
                return 'screen reader info';
              }),
            },
            renderOptions: {},
          },
        );
        const option1 = getByTitle(list[0].title);
        fireEvent.click(option1);

        // it is already selected, should not call callback
        expect(clickHandler).not.toBeCalled();

        const option2 = getByTitle(list[1].title);
        fireEvent.click(option2);

        expect(clickHandler).toBeCalledWith('en');
      });
    });
  });
});
