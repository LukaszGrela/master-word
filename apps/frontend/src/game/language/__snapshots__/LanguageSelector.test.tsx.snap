// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > language > LanguageSelector > Renders with context and all props 1`] = `
<div>
  <div
    class="language-selector translation test"
  >
    <span
      class="hidden sr"
    >
      screenReaderInfo
    </span>
  </div>
</div>
`;

exports[`game > language > LanguageSelector > Renders with context and required props 1`] = `
<div>
  <div
    class="language-selector translation"
  >
    <span
      class="hidden sr"
    >
      screen reader info
    </span>
  </div>
</div>
`;

exports[`game > language > LanguageSelector > Renders with required props 1`] = `
<div>
  <div
    class="language-selector translation"
  >
    <span
      class="hidden sr"
    />
  </div>
</div>
`;

exports[`game > language > LanguageSelector > renders options 1`] = `
<div>
  <div
    class="language-selector translation"
  >
    <span
      class="hidden sr"
    >
      screen reader info
    </span>
    <button
      class="translation-btn"
      title="Option 1 Button Title"
    >
      <span
        class="label"
      >
        Option 1
      </span>
      <span
        class="flag"
      >
        ICO
      </span>
    </button>
    <button
      class="translation-btn"
      title="Option 2 Button Title"
    >
      <span
        class="label"
      >
        Option 2
      </span>
      <span
        class="flag"
      >
        ICO
      </span>
    </button>
  </div>
</div>
`;
